xarray/core/combine.py

```diff
@@ -456,107 +146,372 @@ def _combine_nd(combined_ids, concat_dims, data_vars='all',
         keys, which specify position within the desired final combined result.
     concat_dims : sequence of str
         The dimensions along which the datasets should be concatenated. Must be
-        in order, and the length must match
+        in order, and the length must match the length of the tuples used as
+        keys in combined_ids. If the string is a dimension name then concat
+        along that dimension, if it is None then merge.
 
     Returns
     -------
     combined_ds : xarray.Dataset
     """
 
-    # Perform N-D dimensional concatenation
+    tile_id, ds = list(combined_ids.items())[0]
+
+    n_dims = len(tile_id)
+    if len(concat_dims) != n_dims:
+        raise ValueError("concat_dims has length {} but the datasets "
+                         "passed are nested in a {}-dimensional structure"
+                         .format(str(len(concat_dims)), str(n_dims)))
+
     # Each iteration of this loop reduces the length of the tile_ids tuples
     # by one. It always combines along the first dimension, removing the first
     # element of the tuple
     for concat_dim in concat_dims:
-        combined_ids = _auto_combine_all_along_first_dim(combined_ids,
-                                                         dim=concat_dim,
-                                                         data_vars=data_vars,
-                                                         coords=coords,
-                                                         compat=compat)
+        combined_ids = _combine_all_along_first_dim(combined_ids,
+                                                    dim=concat_dim,
+                                                    data_vars=data_vars,
+                                                    coords=coords,
+                                                    compat=compat)
     combined_ds = list(combined_ids.values())[0]
```

Comment:
At this point, we know there should only be one dataset left, right? In that case, it's often cleaner to use tuple unpacking, because it makes it clear to readers of the code that we aren't throwing anything away, e.g.,
```suggestion
    (combined_ds,) = combined_ids.values()
```